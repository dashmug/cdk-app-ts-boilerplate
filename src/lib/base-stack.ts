import { Stack, StackProps, Tags } from "aws-cdk-lib";
import type { Construct } from "constructs";
import * as packageInfo from "../../package.json";

/**
 * A base stack class with preset environment and tags.
 */
export abstract class BaseStack extends Stack {
  protected constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, {
      env: {
        account: process.env.CDK_DEFAULT_ACCOUNT,
        region: process.env.CDK_DEFAULT_REGION,
      },
      ...props,
    });

    Tags.of(this).add("Application", packageInfo.name);

    if (packageInfo.description) {
      Tags.of(this).add("Description", packageInfo.description);
    }

    if (packageInfo.repository) {
      Tags.of(this).add("Repository", packageInfo.repository);
    }

    Tags.of(this).add("Stack", id);
  }
}
