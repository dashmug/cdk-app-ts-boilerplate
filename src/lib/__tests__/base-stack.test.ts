import { App, StackProps } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { Bucket } from "aws-cdk-lib/aws-s3";
import type { Construct } from "constructs";
import { BaseStack } from "src/lib/base-stack";

class TestStack extends BaseStack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    new Bucket(this, "Bucket");
  }
}

describe("BaseStack", () => {
  const app = new App();
  const stack = new TestStack(app, "TestStack");
  const template = Template.fromStack(stack);

  it("adds resource tags based on package.json values", () => {
    template.hasResourceProperties("AWS::S3::Bucket", {
      Tags: [
        {
          Key: "myorg:application:description",
          Value: "Opinionated CDK Typescript Boilerplate",
        },
        {
          Key: "myorg:application:environment",
          Value: "prod",
        },
        {
          Key: "myorg:application:name",
          Value: "cdk-app-ts-boilerplate",
        },
        {
          Key: "myorg:application:repository",
          Value: "https://github.com/dashmug/cdk-app-ts-boilerplate",
        },
        {
          Key: "myorg:project:customer",
          Value: "myclient",
        },
        {
          Key: "myorg:project:name",
          Value: "myproject",
        },
        {
          Key: "myorg:project:owner",
          Value: "myteam",
        },
        {
          Key: "myorg:stack:name",
          Value: "TestStack",
        },
      ],
    });
  });
});
