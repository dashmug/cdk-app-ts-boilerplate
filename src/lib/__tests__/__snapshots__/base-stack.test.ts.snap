// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseStack generates an expected CloudFormation template 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Bucket83908E77": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "Tags": [
          {
            "Key": "myorg:application:description",
            "Value": "Opinionated CDK Typescript Boilerplate",
          },
          {
            "Key": "myorg:application:environment",
            "Value": "prod",
          },
          {
            "Key": "myorg:application:name",
            "Value": "cdk-app-ts-boilerplate",
          },
          {
            "Key": "myorg:application:repository",
            "Value": "https://github.com/dashmug/cdk-app-ts-boilerplate",
          },
          {
            "Key": "myorg:project:customer",
            "Value": "myclient",
          },
          {
            "Key": "myorg:project:name",
            "Value": "myproject",
          },
          {
            "Key": "myorg:project:owner",
            "Value": "myteam",
          },
          {
            "Key": "myorg:stack:name",
            "Value": "TestStack",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "Queue4A7E3555": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Tags": [
          {
            "Key": "myorg:application:description",
            "Value": "Opinionated CDK Typescript Boilerplate",
          },
          {
            "Key": "myorg:application:environment",
            "Value": "prod",
          },
          {
            "Key": "myorg:application:name",
            "Value": "cdk-app-ts-boilerplate",
          },
          {
            "Key": "myorg:application:repository",
            "Value": "https://github.com/dashmug/cdk-app-ts-boilerplate",
          },
          {
            "Key": "myorg:project:customer",
            "Value": "myclient",
          },
          {
            "Key": "myorg:project:name",
            "Value": "myproject",
          },
          {
            "Key": "myorg:project:owner",
            "Value": "myteam",
          },
          {
            "Key": "myorg:stack:name",
            "Value": "TestStack",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
